core/src/main/java/org/frankframework/pipes/ForPipe.java

```diff
@@ -0,0 +1,169 @@
+ */
+@Forward(name = "stop", description = "exit for loop")
+@Forward(name = "continue", description = "continue in for loop")
+@ElementType(ElementTypes.ITERATOR)
+public class ForPipe extends FixedForwardPipe {
+
+	static final String STOP_FORWARD_NAME = "stop";
+	static final String CONTINUE_FORWARD_NAME = "continue";
+	static final String STOP_AT_PARAMETER_VALUE = "stopAt";
+	private static final String INCREMENT_SESSION_KEY_SUFFIX = "iteration";
+	private @Getter int startAt = 0;
+	private @Getter Integer stopAt = null;
+
+	@Override
+	public void configure() throws ConfigurationException {
+		super.configure();
+
+		// Mandatory stopAt attribute / parameter
+		if (stopAt == null && !getParameterList().hasParameter(STOP_AT_PARAMETER_VALUE)) {
+			throw new ConfigurationException("Value for 'stopAt' is mandatory to break out of the for loop pipe");
+		}
+
+		if (stopAt != null && getParameterList().hasParameter(STOP_AT_PARAMETER_VALUE)) {
+			throw new ConfigurationException("Value for 'stopAt' is set both as attribute and Parameter, please use one of the two options");
+		}
+
+		// Mandatory forwards
+		PipeForward stopForward = findForward(STOP_FORWARD_NAME);
+		PipeForward continueForward = findForward(CONTINUE_FORWARD_NAME);
+
+		if (stopForward == null) {
+			throw new ConfigurationException("has no forward with name [" + STOP_FORWARD_NAME + "]");
+		}
+
+		if (continueForward == null) {
+			throw new ConfigurationException("has no forward with name [" + CONTINUE_FORWARD_NAME + "]");
+		}
+	}
+
+	@Override
+	public PipeRunResult doPipe(Message message, PipeLineSession session) throws PipeRunException {
+		Integer stopAtValue = determineStopAtValue(message, session);
+		String sessionKey = getIncrementSessionKey();
+
+		// Get the value from the session, default to the startAt value
+		int i = session.get(sessionKey, startAt);
+
+		// If i < stopAtValue, increment and continue
+		if (i < stopAtValue) {
+			PipeForward continueForward = findForward(CONTINUE_FORWARD_NAME);
+
+			session.put(sessionKey, ++i);
+
+			return new PipeRunResult(continueForward, message);
+		}
+
+		// Else, remove the session key and forward to the stop forward
+		session.remove(sessionKey);
+		return new PipeRunResult(findForward(STOP_FORWARD_NAME), message);
+	}
+
+	String getIncrementSessionKey() {
+		return String.format("%s.%s", getName(), INCREMENT_SESSION_KEY_SUFFIX);
+	}
+
+	private Integer determineStopAtValue(Message message, PipeLineSession session) throws PipeRunException {
+		if (stopAt != null) {
+			return stopAt;
+		}
+
+		Optional<ParameterValueList> optionalParameterValueList = getParameterValueList(message, session);
+
+		if (optionalParameterValueList.isPresent()) {
+			ParameterValue stopAtParameter = optionalParameterValueList.get().get(STOP_AT_PARAMETER_VALUE);
+
+			if (stopAtParameter.getValue() != null && StringUtils.isNotBlank(stopAtParameter.asStringValue())) {
+				return stopAtParameter.asIntegerValue(0);
+			}
```

Comment:
Je had ook deze if weg kunnen laten en
`return stopAtParameter.asIntegerValue(attributeWaarde);`
kunnen gebruiken. Een parameter is overigens nooit null dacht ik?

Reply:
Ok, ik check dit nog even